gapii/cc/spy.cpp

```diff
@@ -239,18 +239,18 @@ Spy::~Spy() {
 
 CallObserver* Spy::enter(const char* name, uint32_t api) {
   lock();
-  auto ctx = new CallObserver(this, gContext, api);
-  ctx->setCurrentCommandName(name);
-  gContext = ctx;
-  return ctx;
+  // This code assumes we only deal with one 'api' in the trace (that is the
+  // case in AGI, which only handles Vulkan): it creates a fresh observer
+  // (per-thread, since tl_callObserver is thread_local) for that API the first
+  // time and reuses it in subsequent calls.
+  if (tl_callObserver == nullptr) {
+    tl_callObserver = new CallObserver(this, api);
```

Comment:
I don't see a corresponding `delete` for this `new`. Is it done implicetly?

Reply:
I noticed that too. There is no implicit delete for the persistent CallObserver. The destructor won't be called, but the memory will be freed on application termination. This was ok for me, but it could actually lead to a problematic memory leak if threads are created frequently. So I propose to use a std::unique_ptr to store the CallObserver.